plugins/inputs/vsphere/vsphere.go

```diff
@@ -0,0 +1,290 @@
+		"cpu.used.summation",
+		"cpu.usage.average",
+		"cpu.utilization.average",
+		"cpu.wait.summation.",
+		"cpu.idle.summation",
+		"cpu.readiness.average",
+		"cpu.costop.summation",
+		"cpu.swapwait.summation",
+		"mem.swapoutRate.average",
+		"disk.kernelReadLatency.average",
+		"disk.kernelWriteLatency.average" ]
+
+# host_metric_exclude = [] ## Nothing excluded by default
+
+############### Clusters ###############
+
+## Collect cluster instance metrics, such as individual cores? (default=true)
+#cluster_instances = true
+
+## Typical cluster metrics (if omitted, all metrics are collected)
+cluster_metric_include = [
+	  "cpu.usage.*",
+	  "cpu.usagemhz.*",
+	  "mem.usage.*",
+	  "mem.active.*" ]
+
+# cluster_metric_exclude = [] ## Nothing excluded by default
+
+############### Datastores ###############
+
+## Collect datastore instance metrics, such as individual LUNs and datafiles? (default=false)
+#datastore_instances = false
+
+## Typical datastore metrics (if omitted, all metrics are collected)
+##
+## WARNING: It is not recommended to include all metrics. Your collection will probably time out if you do!
+##
+datastore_metric_include = [
+	"datastore.read.average",
+	"datastore.write.average",
+	"datastore.throughput.usage.average",	
+	"datastore.throughput.contention.average" ]
+
+## Number of objects to retreive per query. 
+# objects_per_query = 256
+
+## Number of metrics (data points) to retrieve in each query for non-realtime metrics (cluster and datastore).
+## Set to 64 for vCenter 5.5 and 6.0 (default: 256)
+# metrics_per_query = 256
+
+## the interval before (re)discovering objects subject to metrics collection (default: 300s)
+# object_discovery_interval = "300s"
+
+## timeout applies to any of the connection request made to vcenter
+# timeout = "20s"
+
+## Optional SSL Config
+# ssl_ca = "/path/to/cafile"
+# ssl_cert = "/path/to/certfile"
+# ssl_key = "/path/to/keyfile"
+## Use SSL but skip chain & host verification
+# insecure_skip_verify = false 
+`
+
+// SampleConfig returns a set of default configuration to be used as a boilerplate when setting up
+// Telegraf.
+func (v *VSphere) SampleConfig() string {
+	return sampleConfig
+}
+
+// Description returns a short textual description of the plugin
+func (v *VSphere) Description() string {
+	return "Read metrics from VMware vCenter"
+}
+
+func (v *VSphere) checkEndpoints() error {
+	if v.endpoints != nil {
+		return nil
+	}
+
+	v.endpoints = make([]*Endpoint, len(v.Vcenters))
+	for i, rawURL := range v.Vcenters {
+		u, err := soap.ParseURL(rawURL)
+		if err != nil {
+			return err
+		}
+
+		v.endpoints[i] = NewEndpoint(v, u)
+	}
+	return nil
+}
+
+// Start is called from telegraf core when a plugin is started and allows it to
+// perform initialization tasks.
+func (v *VSphere) Start(acc telegraf.Accumulator) error {
+	log.Println("D! [input.vsphere]: Starting plugin")
+	rootCtx, cancel := context.WithCancel(context.Background())
+	v.rootCtx = rootCtx
```

Comment:
I know that storing a context is frowned upon, but it works perfectly in this case and it's very hard to solve in any other way. The issue is the background discovery goroutine. Even if Stop() isn't called until Gather() has finished, you may still have a discovery running. Cancelling its context would immediately stop the discovery, even if it's stuck on SOAP calls, since the same context would be passed all the way down. If you have a better solution for the discover issue, I'd be glad to make the change. The stored context is sticking out like a sore thumb to me too!

Reply:
Nevermind... I think I have a solution!

Reply:
Done. I still need to hold on to the CancelFunc for the context I use for discovery, but it's passed, not stored. Also, I still need something to wait on to make sure all outstanding discovery is finished/aborted before I return from Stop(), but since we're only waiting for discovery now, I can solve that using a simple Mutex rather than that funky ConcurrentWaitgroup thing.